File: FruitIntoBaskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        if(n < 2)
            return 1;
        int fruit1 = fruits[0];
        int fruit2 = fruits[0];
        int sum = 0;
        
        
        int a = 0 , b = 1, c = 0;
        
        while(b < n)
        {
            if(fruit1 == fruit2)
            {
                c = b;
                fruit2 = fruits[b];
            }
            
            if(fruits[b] == fruit1 || fruits[b] == fruit2)
            {
                if(fruits[b] != fruits[c])
                    c = b;
                sum = max(b-a+1,sum);
            }
            else
            {
                a = c ;
                c = b;
                fruit1 = fruits[a];
                fruit2 = fruits[b];
            }
            b++;
        }
        
        return sum;
    }
};

File: FruitIntoBaskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:

        left = 0
        maxFruits = 1
        n = len(fruits)
        seen = defaultdict(int)

        for right in range(n):

            seen[fruits[right]] += 1

            while left <= right and len(seen) > 2:
                seen[fruits[left]] -= 1
                if seen[fruits[left]] == 0:
                    del seen[fruits[left]]

                left += 1
            
            maxFruits = max(maxFruits, right - left + 1)

        return maxFruits
